### PR TITLE
Feat/clear bad debt

### LINKED ISSUE #811: Bad debt repayment proposal & code

### PR DESCRIPTION
**Pull request by xklob**
Provides a contract, `FuseFixer.sol`, that gives capabilities of paying back bad debt specifically for the exploiter.
The bad debt can be repaid over multiple transactions, or all at once.

Is not opinionated about how the contract gets its assets.

Todo: need to add tests for the "multiple transactions" path.
Todo: need to think about how this could be griefed, and preventions we might want to implement.

(note that we have a slippage parameter of sorts in two ways: first, by not sending too many assets to this contract; second, by using the individual-repay method and providing a maximum amount)

### CODE REVIEW

File: contracts/utils/FuseFixer.sol [hunk 1]
```diff
@@ -0,0 +1,182 @@
...
+import "@openzeppelin/contracts/token/ERC20/utils/SafeERC20.sol";
+import {CoreRef} from "../refs/CoreRef.sol";
+import {TribeRoles} from "../core/TribeRoles.sol";
+
+/// @title base class for a Compound PCV Deposit
+/// @author Fei Protocol
+contract FuseFixer is PCVDeposit {
+    address constant DEBTOR = address(0x32075bAd9050d4767018084F0Cb87b3182D36C45);
+
+    address[] internal UNDERLYINGS = [
```
Comment (ElliotFriedman) [inline contracts/utils/FuseFixer.sol]: can this be made immutable or constant?

File: contracts/utils/FuseFixer.sol [hunk 2]
```diff
@@ -0,0 +1,182 @@
...
+        address(0x03ab458634910AaD20eF5f1C8ee96F1D6ac54919), // RAI
+        address(0x6B175474E89094C44Da98b954EedeAC495271d0F), // DAI
+        address(0xA0b86991c6218b36c1d19D4a2e9Eb0cE3606eB48), // USDC
+        address(0x5f98805A4E8be255a32880FDeC7F6728C6568bA0), // LUSD
+        address(0x7f39C581F595B53c5cb19bD0b3f8dA6c935E2Ca0), // wstETH
+        address(0xa693B19d2931d498c5B318dF961919BB4aee87a5), // USTw
+        address(0xdAC17F958D2ee523a2206206994597C13D831ec7) // USDT
+    ];
+
+    address[] internal CTOKENS = [
```
Comment (ElliotFriedman) [inline contracts/utils/FuseFixer.sol]: can this be made immutable or constant?

File: contracts/utils/FuseFixer.sol [hunk 3]
```diff
@@ -0,0 +1,182 @@
...
+            _repayETH();
+        } else {
+            _repayERC20(underlying);
+        }
+    }
+
+    /* Helper Functions */
+
+    function _buildCTokenMapping() internal {
+        for (uint256 i = 0; i < CTOKENS.length; i++) {
```
Comment (ElliotFriedman) [inline contracts/utils/FuseFixer.sol:132]: gas opti: cache CTOKENS.length & make increment unchecked

File: contracts/utils/FuseFixer.sol [hunk 4]
```diff
@@ -0,0 +1,182 @@
...
+            address token = CTOKENS[i];
+            address underlying = CTokenFuse(token).underlying();
+            underlyingsToCTokens[underlying].push(token);
+        }
+    }
+
+    function _approveCTokens() internal {
+        // UNDERLYINGS[0] == ETH, so we skip that one
+        for (uint256 i = 1; i < UNDERLYINGS.length; i++) {
```
Comment (ElliotFriedman) [inline contracts/utils/FuseFixer.sol]: gas opti: cache UNDERLYINGS.length & make increment unchecked

File: contracts/utils/FuseFixer.sol [hunk 5]
```diff
@@ -0,0 +1,182 @@
...
+            address underlying = UNDERLYINGS[i];
+            address[] memory ctokens = underlyingsToCTokens[underlying];
+
+            for (uint256 j = 0; j < ctokens.length; j++) {
```
Comment (ElliotFriedman) [inline contracts/utils/FuseFixer.sol:149]: gas opti: cache ctokens.length & make increment unchecked

File: contracts/utils/FuseFixer.sol [hunk 6]
```diff
@@ -0,0 +1,182 @@
...
+                address ctoken = ctokens[j];
+                SafeERC20.safeApprove(IERC20(underlying), ctoken, type(uint256).max);
+            }
+        }
+    }
+
+    function _repayETH() internal {
+        address[] memory cEtherTokens = underlyingsToCTokens[address(0)];
+
+        for (uint256 i = 0; i < cEtherTokens.length; i++) {
```
Comment (ElliotFriedman) [inline contracts/utils/FuseFixer.sol:160]: gas opti: cache cEtherTokens.length & make increment unchecked

File: contracts/utils/FuseFixer.sol [hunk 7]
```diff
@@ -0,0 +1,182 @@
...
+            CEtherFuse token = CEtherFuse(cEtherTokens[i]);
+            uint256 debtAmount = token.borrowBalanceCurrent(DEBTOR);
+            token.repayBorrowBehalf{value: debtAmount}(DEBTOR);
+        }
+    }
+
+    function _repayERC20(address underlying) internal {
+        address[] memory cERC20Tokens = underlyingsToCTokens[underlying];
+
+        for (uint256 i = 0; i < cERC20Tokens.length; i++) {
```
Comment (ElliotFriedman) [inline contracts/utils/FuseFixer.sol:173]: gas opti: cache cERC20Tokens.length & make increment unchecked

File: contracts/utils/FuseFixer.sol [hunk 8]
```diff
@@ -0,0 +1,182 @@
...
+import {PCVDeposit} from "../pcv/PCVDeposit.sol";
+import {CTokenFuse, CEtherFuse} from "../external/fuse/CToken.sol";
+import "@openzeppelin/contracts/token/ERC20/IERC20.sol";
+import "@openzeppelin/contracts/token/ERC20/utils/SafeERC20.sol";
+import {CoreRef} from "../refs/CoreRef.sol";
+import {TribeRoles} from "../core/TribeRoles.sol";
+
+/// @title base class for a Compound PCV Deposit
+/// @author Fei Protocol
+contract FuseFixer is PCVDeposit {
```
Comment (ElliotFriedman) [inline contracts/utils/FuseFixer.sol:13]: Do we need this contract to be a PCVDeposit? I don't see it added to the CR oracle in the DAO script.

File: contracts/utils/FuseFixer.sol (hunk 1 above)
Comment (xklob) [reply contracts/utils/FuseFixer.sol]: Nope, I did try. Arrays can't be immutable or constant.

File: contracts/utils/FuseFixer.sol (hunk 2 above)
Comment (xklob) [reply contracts/utils/FuseFixer.sol]: (see above)

File: contracts/utils/FuseFixer.sol (hunk 3 above)
Comment (xklob) [reply contracts/utils/FuseFixer.sol:132]: Normally would agree but want to leave this as readable and easy to check for errors as possible. It's a one-off anyways.

File: contracts/utils/FuseFixer.sol (hunk 4 above)
Comment (xklob) [reply contracts/utils/FuseFixer.sol]: (see above)

File: contracts/utils/FuseFixer.sol (hunk 5 above)
Comment (xklob) [reply contracts/utils/FuseFixer.sol:149]: (see above)

File: contracts/utils/FuseFixer.sol (hunk 6 above)
Comment (xklob) [reply contracts/utils/FuseFixer.sol:160]: (see above)

File: contracts/utils/FuseFixer.sol (hunk 7 above)
Comment (xklob) [reply contracts/utils/FuseFixer.sol:173]: (see above)

File: contracts/utils/FuseFixer.sol (hunk 8 above)
Comment (xklob) [reply contracts/utils/FuseFixer.sol:13]: It's a TC script, not a DAO script, so we can't add it to the CR oracle anyways.

Either way, assets in this contract are slated for destruction, as it were, so it doesn't make sense to count them in PCV anyway.

It should be a PCV deposit however so that the PCVGuardian can withdraw the remaining funds (what little should remain) to a safe deposit.

File: contracts/utils/FuseFixer.sol [hunk 9]
```diff
@@ -0,0 +1,189 @@
...
+        assert(underlyingsToCTokens[UNDERLYINGS[7]].length == 2);
+        assert(underlyingsToCTokens[UNDERLYINGS[8]].length == 4);
+        assert(underlyingsToCTokens[UNDERLYINGS[9]].length == 1);
+
+        // send out token approvals
+        _approveCTokens();
+    }
+
+    /// @dev Repay all debt
+    function repayAll() public onlyTribeRole(TribeRoles.PCV_GUARDIAN_ADMIN) {
```
Comment (xklob) [inline contracts/utils/FuseFixer.sol]: Is the PCV Guardian Admin the appropriate role here? If not, what should be? Was a bit stumped by our new role system.

File: contracts/utils/FuseFixer.sol [hunk 10]
```diff
@@ -0,0 +1,189 @@
+// SPDX-License-Identifier: GPL-3.0-or-later
+pragma solidity ^0.8.0;
+
+import "@openzeppelin/contracts/token/ERC20/IERC20.sol";
```
Comment (thomas-waite) [inline contracts/utils/FuseFixer.sol]: To be consistent, would use `import {} from "@openzeppelin/" syntax for these imports too

File: contracts/utils/FuseFixer.sol [hunk 11]
```diff
@@ -0,0 +1,189 @@
...
+/// @author Fei Protocol
+contract FuseFixer is PCVDeposit {
+    address constant DEBTOR = address(0x32075bAd9050d4767018084F0Cb87b3182D36C45);
+
+    address[] internal UNDERLYINGS = [
+        address(0x0000000000000000000000000000000000000000), // ETH
+        address(0x956F47F50A910163D8BF957Cf5846D573E7f87CA), // FEI
+        address(0x853d955aCEf822Db058eb8505911ED77F175b99e), // FRAX
+        address(0x03ab458634910AaD20eF5f1C8ee96F1D6ac54919), // RAI
+        address(0x6B175474E89094C44Da98b954EedeAC495271d0F), // DAI
```
Comment (thomas-waite) [inline contracts/utils/FuseFixer.sol:21]: Minor, but do these need explicit casting to an address? Surely defining `address[] internal UNDERLYINGS = []` tells the compiler that they are addresses?

File: contracts/utils/FuseFixer.sol [hunk 12]
```diff
@@ -0,0 +1,189 @@
...
+        address(0xA0b86991c6218b36c1d19D4a2e9Eb0cE3606eB48), // USDC
+        address(0x5f98805A4E8be255a32880FDeC7F6728C6568bA0), // LUSD
+        address(0x7f39C581F595B53c5cb19bD0b3f8dA6c935E2Ca0), // wstETH
+        address(0xa693B19d2931d498c5B318dF961919BB4aee87a5), // USTw
+        address(0xdAC17F958D2ee523a2206206994597C13D831ec7) // USDT
```
Comment (thomas-waite) [inline contracts/utils/FuseFixer.sol:26]: Was `WETH` not also an affected asset? It's in the report

File: contracts/utils/FuseFixer.sol [hunk 13]
```diff
@@ -0,0 +1,189 @@
...
+        for (uint256 i = 0; i < cEtherTokens.length; i++) {
+            CEtherFuse token = CEtherFuse(cEtherTokens[i]);
+            uint256 debtAmount = token.borrowBalanceCurrent(DEBTOR);
+            if (debtAmount > 0) {
+                token.repayBorrowBehalf{value: debtAmount}(DEBTOR);
+            }
+        }
+    }
+
+    function _repayERC20(address underlying) internal {
```
Comment (thomas-waite) [inline contracts/utils/FuseFixer.sol:170]: Even though it's a one time use contract, I'd still add natspec comments to make reviewing easier

File: contracts/utils/FuseFixer.sol [hunk 14]
```diff
@@ -0,0 +1,189 @@
...
+        for (uint256 i = 0; i < CTOKENS.length; i++) {
+            address token = CTOKENS[i];
+            address underlying = CTokenFuse(token).underlying();
+            underlyingsToCTokens[underlying].push(token);
+        }
+    }
+
+    function _approveCTokens() internal {
+        // UNDERLYINGS[0] == ETH, so we skip that one
+        for (uint256 i = 1; i < UNDERLYINGS.length; i++) {
```
Comment (thomas-waite) [inline contracts/utils/FuseFixer.sol]: I get why you do this, but feel it's leaving a footgun - if you change the `UNDERLYINGS` order around you'll skip the wrong `cToken`. 

Would personally check for `underlying != address(0x0)` or similar to be more explicit. One time use, so no strong feelings

File: contracts/utils/FuseFixer.sol [hunk 15]
```diff
@@ -0,0 +1,189 @@
...
+
+        for (uint256 i = 0; i < cERC20Tokens.length; i++) {
+            CTokenFuse token = CTokenFuse(cERC20Tokens[i]);
+            uint256 debtAmount = token.borrowBalanceCurrent(DEBTOR);
+            token.repayBorrowBehalf(DEBTOR, debtAmount);
+        }
+    }
+
+    // Must be non-view since this updates a state var
+    function getTotalDebt(address underlying) internal returns (uint256 debt) {
```
Comment (thomas-waite) [inline contracts/utils/FuseFixer.sol]: Worthwhile making this `public`? Can imagine it will be helpful for off-chain analysis as we will likely pay back asset by asset

File: contracts/utils/FuseFixer.sol (hunk 9 above)
Comment (thomas-waite) [inline contracts/utils/FuseFixer.sol]: Think this is the plan anyway, but much more in favour of partially repaying and checking progress/status as we go

File: proposals/dao/repay_fuse_bad_debt.ts [hunk 16]
```diff
@@ -0,0 +1,172 @@
...
+/*
+Withdraw FEI from Aave and Compound
+*/
+
+const fipNumber = 'repay_fuse_bad_debt'; // Change me!
+
+// Do any deployments
+// This should exclusively include new contract deployments
+const deploy: DeployUpgradeFunc = async (deployAddress: string, addresses: NamedAddresses, logging: boolean) => {
+  const fuseFixerFactory = (await hre.ethers.getContractFactory('FuseFixer')) as FuseFixer__factory;
```
Comment (thomas-waite) [inline proposals/dao/repay_fuse_bad_debt.ts:23]: Typescript should be able to infer the `fuseFixerFactory` type without explicitly using `as FuseFixer_factory` I believe

File: proposals/dao/repay_fuse_bad_debt.ts [hunk 17]
```diff
@@ -0,0 +1,172 @@
...
+  const fuseFixer = await fuseFixerFactory.deploy(addresses.core);
```
Comment (thomas-waite) [inline proposals/dao/repay_fuse_bad_debt.ts:24]: `.deploy()` just sends the `deploy` transaction. Good practice to wait for that to be mined with `await fuseFixer.deployed()`

File: proposals/dao/repay_fuse_bad_debt.ts [hunk 18]
```diff
@@ -0,0 +1,172 @@
...
+    '0xd632f22692fac7611d2aa1c0d552930d43caed3b',
+    '0xc9BC48c72154ef3e5425641a3c747242112a46AF',
+    '0x5d3a536E4D6DbD6114cc1Ead35777bAB948E3643',
+    '0x0A59649758aa4d66E25f08Dd01271e891fe52199',
+    '0x66017d22b0f8556afdd19fc67041899eb65a21bb',
+    '0x2faf487a4414fe77e2327f0bf4ae2a264a776ad2',
+    '0x5754284f345afc66a98fbb0a0afe71e0f007b949'
+  ]);
+
+  const fei = await hre.ethers.getContractAt(
```
Comment (thomas-waite) [inline proposals/dao/repay_fuse_bad_debt.ts:60]: Not a fan of these various hardcoded addresses. Would personally put the various tokens into `mainnetAddresses` with an `External` category and then do something like 

`await contracts.frax.connect(signer).transfer()`

File: proposals/dao/repay_fuse_bad_debt.ts [hunk 19]
```diff
@@ -0,0 +1,172 @@
...
+  ];
+
+  for (const ctokenAddress of ctokens) {
+    const ctoken = await hre.ethers.getContractAt('CTokenFuse', ctokenAddress);
+    const debt = await ctoken.callStatic.borrowBalanceCurrent(debtor);
+
+    if (debt.gt(0)) {
+      throw new Error(`Debt for ctoken ${ctokenAddress} is greater than 0: ${debt}`);
+    }
+  }
```
Comment (thomas-waite) [inline proposals/dao/repay_fuse_bad_debt.ts:172]: I would probably make that `fuseFixer.getTotalDebt()` function on the contract public and assert it returns 0 here too

Comment (thomas-waite) [reply proposals/dao/repay_fuse_bad_debt.ts:172]: Maybe have it emit an event with the total debt and catch that event + inspect the value

**Review by thomas-waite — COMMENTED** [review]
Will have a think about attack vectors and edge cases, only have small comments so far. Added suggestion for the role

Main thing as you say is the tests for multiple transactions. I would probably break those out into an actual e2e test file rather than just do in the `validate()` function

File: contracts/utils/FuseFixer.sol (hunk 9 above)
Comment (thomas-waite) [reply contracts/utils/FuseFixer.sol]: Would set this to `PCV_SAFE_MOVER_ROLE` - that's the role on the guardian that allows an address to move PCV funds to a Safe address. The `tribalCouncilTimelock` has it and is the only address with it. Think probably the most appropriate, particularly of the roles the `tribalCouncilTimelock` has.

`PCV_GUARDIAN_ADMIN` just allows safe addresses to be set and unset, so not appropriate here imo. The `optimisticTimelock` also has `PCV_GUARDIAN_ADMIN` currently, probably don't want that to be able to trigger a repayment

Comment (Joeysantoro) [reply contracts/utils/FuseFixer.sol]: Agree with safe mover, as this is likely the role that will be getting assets into this contract in the first place

File: contracts/utils/FuseFixer.sol (hunk 12 above)
Comment (Joeysantoro) [reply contracts/utils/FuseFixer.sol:26]: technically its native ETH

File: contracts/utils/FuseFixer.sol (hunk 9 above)
Comment (xklob) [reply contracts/utils/FuseFixer.sol]: Noted, will update to `PCV_SAFE_MOVE_ROLE`

Comment (xklob) [reply contracts/utils/FuseFixer.sol]: Updated.

File: contracts/utils/FuseFixer.sol (hunk 11 above)
Comment (xklob) [reply contracts/utils/FuseFixer.sol:21]: Great question - this is the way I've always done it. But perhaps not!

File: contracts/utils/FuseFixer.sol (hunk 13 above)
Comment (xklob) [reply contracts/utils/FuseFixer.sol:170]: Natspec isn't parsed for internal functions

Comment (xklob) [reply contracts/utils/FuseFixer.sol:170]: But will add some comments either way

File: contracts/utils/FuseFixer.sol (hunk 14 above)
Comment (xklob) [reply contracts/utils/FuseFixer.sol]: Done.

File: contracts/utils/FuseFixer.sol (hunk 15 above)
Comment (xklob) [reply contracts/utils/FuseFixer.sol]: Done.

File: proposals/dao/repay_fuse_bad_debt.ts (hunk 16 above)
Comment (xklob) [reply proposals/dao/repay_fuse_bad_debt.ts:23]: Not in this format, apparently. Just returns contract.

File: proposals/dao/repay_fuse_bad_debt.ts (hunk 17 above)
Comment (xklob) [reply proposals/dao/repay_fuse_bad_debt.ts:24]: Great catch

File: proposals/dao/repay_fuse_bad_debt.ts (hunk 18 above)
Comment (xklob) [reply proposals/dao/repay_fuse_bad_debt.ts:60]: If this was code that was going to remain in use, I'd agree. But the # of tokens here and the fact that this is a one-off - I think it's okay.

File: proposals/dao/repay_fuse_bad_debt.ts (hunk 19 above)
Comment (xklob) [reply proposals/dao/repay_fuse_bad_debt.ts:172]: Great call

File: contracts/utils/FuseFixer.sol [hunk 20]
```diff
@@ -0,0 +1,200 @@
...
+pragma solidity ^0.8.0;
+
+import {IERC20} from "@openzeppelin/contracts/token/ERC20/IERC20.sol";
+import {SafeERC20} from "@openzeppelin/contracts/token/ERC20/utils/SafeERC20.sol";
+import {PCVDeposit} from "../pcv/PCVDeposit.sol";
+import {CTokenFuse, CEtherFuse} from "../external/fuse/CToken.sol";
+import {CoreRef} from "../refs/CoreRef.sol";
+import {TribeRoles} from "../core/TribeRoles.sol";
+
+/// @title base class for a Compound PCV Deposit
```
Comment (eswak) [inline contracts/utils/FuseFixer.sol]: nit

File: contracts/utils/FuseFixer.sol [hunk 21]
```diff
@@ -0,0 +1,200 @@
...
+    /// @notice This is not a view function! Use .staticcall in ethers to get the return value
+    /// @param underlying the asset to get the debt for; pass in 0x0 for ETH
+    /// @return debt the total debt for the asset
+    function getTotalDebt(address underlying) public returns (uint256 debt) {
+        for (uint256 i = 0; i < CTOKENS.length; i++) {
+            CTokenFuse token = CTokenFuse(CTOKENS[i]);
+            if (token.underlying() == underlying) {
+                debt += CTokenFuse(CTOKENS[i]).borrowBalanceCurrent(DEBTOR);
+            }
+        }
```
Comment (Joeysantoro) [inline contracts/utils/FuseFixer.sol:125]: couldn't you just iterate over the underlyingsToCtokens map ctokens rather than the whole CTokens. Then underlying is always =.

File: proposals/description/repay_fuse_bad_debt.ts [hunk 22]
```diff
@@ -19,6 +19,13 @@ export default fip_x;*/
 const fip_x: ProposalDescription = {
   title: 'Repay Fuse Bad Debt',
   commands: [
+    {
+      target: 'pcvGuardianNew',
+      values: '0',
+      method: 'setSafeAddress(address)',
+      arguments: ['{fuseFixer}'],
```
Comment (thomas-waite) [inline proposals/description/repay_fuse_bad_debt.ts:26]: If we're making this a safe address, think we should add a generic `withdrawERC20()` function to. In case funds are accidentally sent to this address, otherwise we could trap them

**Review by thomas-waite — COMMENTED** [review]
My feedback was addressed + have reviewed the subsequently pushed commits

File: contracts/utils/FuseFixer.sol [hunk 23]
```diff
@@ -0,0 +1,200 @@
...
+        for (uint256 i = 1; i < UNDERLYINGS.length; i++) {
+            _repayERC20(UNDERLYINGS[i]);
+        }
+    }
+
+    /// @dev Repay the underlying asset on all ctokens up to the maximum provide
+    /// @notice reverts if the total bad debt is beyond the provided maximum
+    /// @param underlying the asset to repay in
+    /// @param maximum the maximum amount of underlying asset to repay
+    function repay(address underlying, uint256 maximum) public onlyTribeRole(TribeRoles.PCV_SAFE_MOVER_ROLE) {
```
Comment (Joeysantoro) [inline contracts/utils/FuseFixer.sol]: add GUARDAIN

File: contracts/utils/FuseFixer.sol [hunk 24]
```diff
@@ -0,0 +1,200 @@
...
+        assert(underlyingsToCTokens[UNDERLYINGS[7]].length == 2);
+        assert(underlyingsToCTokens[UNDERLYINGS[8]].length == 4);
+        assert(underlyingsToCTokens[UNDERLYINGS[9]].length == 1);
+
+        // send out token approvals
+        _approveCTokens();
+    }
+
+    /// @dev Repay all debt
+    function repayAll() public onlyTribeRole(TribeRoles.PCV_SAFE_MOVER_ROLE) {
```
Comment (Joeysantoro) [inline contracts/utils/FuseFixer.sol]: add GUARDIAN

File: proposals/description/repay_fuse_bad_debt.ts [hunk 25]
```diff
@@ -0,0 +1,96 @@
...
+  title: 'Repay Fuse Bad Debt',
+  commands: [
+    {
+      target: 'pcvGuardianNew',
+      values: '0',
+      method: 'setSafeAddress(address)',
+      arguments: ['{fuseFixer}'],
+      description: 'Set FuseFixer as a safe address'
+    },
+    {
```
Comment (Joeysantoro) [inline proposals/description/repay_fuse_bad_debt.ts:34]: imo all the repays should be taken out. The PCV Guardian can safely repay one by one. This proposal should simply be setSafeAddress on fuseFixed imo

File: contracts/utils/FuseFixer.sol (hunk 21 above)
Comment (xklob) [reply contracts/utils/FuseFixer.sol:125]: That would be more gas efficient, yes.

File: proposals/description/repay_fuse_bad_debt.ts (hunk 25 above)
Comment (xklob) [reply proposals/description/repay_fuse_bad_debt.ts:34]: Sure, this was just for testing purposes.

File: contracts/utils/FuseFixer.sol (hunk 21 above)
Comment (xklob) [reply contracts/utils/FuseFixer.sol:125]: Leaving as-is since it works and this is a one-off.

File: proposals/description/repay_fuse_bad_debt.ts (hunk 25 above)
Comment (xklob) [reply proposals/description/repay_fuse_bad_debt.ts:34]: Done.

**Comment by xklob** [comment]
Deployed to `0xFE7547F583aAe1212e72e063Aac25057C06c4797`

**Comment by xklob** [comment]
Verified on etherscan: https://etherscan.io/address/0xfe7547f583aae1212e72e063aac25057c06c4797#code

**Review by Joeysantoro — APPROVED** [review]
